### PR TITLE
Generate AWS Client Configuration Defaults

### PR DESCRIPTION
**Pull request by skmcgrail**
Generates default settings for Smithy clients for HTTP Signing, Retryer, HTTP Client, and Idempotency Token Provider. This also includes changes to the SigV4 middleware (signing interface and configuration).

Requires https://github.com/awslabs/smithy-go/pull/152

### CODE REVIEW

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsClientUserAgent.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    public void writeAdditionalFiles(
+            GoSettings settings, Model model, SymbolProvider symbolProvider, GoDelegator goDelegator
+    ) {
+        ServiceTrait serviceTrait = settings.getService(model).expectTrait(ServiceTrait.class);
+
+        goDelegator.useShapeWriter(settings.getService(model), writer -> {
+            writer.openBlock("func $L(stack $P) {", "}", MIDDLEWARE_RESOLVER, SymbolUtils.createPointableSymbolBuilder("Stack",
+                    SmithyGoDependency.SMITHY_MIDDLEWARE).build(), () -> {
+                // TODO: This should include the module version
+                writer.write("$T($S)(stack)", SymbolUtils.createValueSymbolBuilder("AddUserAgentKey",
```
Comment (jasdel) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsClientUserAgent.java:46]: Probably want a github issue for tracking this

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsSdkClientId.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+package software.amazon.smithy.aws.go.codegen;
+
+import software.amazon.smithy.aws.traits.ServiceTrait;
+import software.amazon.smithy.go.codegen.GoSettings;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.ServiceShape;
+
+public class AwsSdkClientId implements GoIntegration {
```
Comment (jasdel) [inline codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsSdkClientId.java:24]: it probably should be `AwsSdkServiceId`

Comment (skmcgrail) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsSdkClientId.java:24]: Good catch!

File: codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsClientUserAgent.java (hunk 1 above)
Comment (skmcgrail) [reply codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/AwsClientUserAgent.java:46]: Created https://github.com/aws/aws-sdk-go-v2/issues/698

**Pull request merged**